saleem-latif/WL-299: SCSS Overrides for Comprehensive Theming

saleem-latif opened the pull request:
Hi @mattdrayer , @andy-armstrong, @mtyaka , @ziafazal , @benpatterson 

Kindly review this PR and share your thoughts on these changes for [WL-299](https://openedx.atlassian.net/browse/WL-299). 

__Note:__
Since, sass first looks at its current path to resolve `@import` so, We decided to move sass files, that comprehensive theme will be able to override, to partials directory. I have only moved  `base/_variables.scss` to partials dir, if this change set looks good than also share your thoughts on what other sass files should be moved to partials dir.

__Description of [WL-299] (https://openedx.atlassian.net/browse/WL-299):__
Currently the behavior of static files and templates overriding in comprehensive theming is that if we place a file named same as it is placed in lms it will override the lms static/template file.
We want to apply the same override pattern for sass files, see the discussion here and here for some insight.
__Acceptance Criteria:__

1. If I place a file named base/_variables.scss inside edx-platform/themes/stanford/lms/static/sass/partials/ than it should override all the definitions from base/_variables.scss present in edx-platform/lms/static/sass/partials/

for more information regarding implementation details look at the comments on [WL-299] (https://openedx.atlassian.net/browse/WL-299)

mattdrayer commented:
:+1: LGTM @saleem-latif -- would be great if we could get one more thumb from the guys you've tagged -- @ziafazal is out for the rest of the week -- perhaps @mtyaka or @andy-armstrong can take a look?

mattdrayer commented:
@AlasdairSwan I'm not sure if you're doing much with the Sass pipeline, but if so maybe you could take a look as well

mtyaka commented on pavelib/assets.py:
```diff
@@ -56,17 +70,21 @@ def configure_paths():
         if lms_sass.isdir():
             lms_css.mkdir_p()
-            THEME_SASS_DIRECTORIES.append(lms_sass)
+            THEME_SASS_DIRECTORIES.append(("lms/static/sass", theme_dir / "lms" / "static" / "css"))
```
Nit: You can replace `theme_dir / "lms" / "static" / "css"` with `lms_css`, which is already defined a couple of lines above.

mtyaka commented on pavelib/assets.py:
```diff
@@ -56,17 +70,21 @@ def configure_paths():
         if cms_sass.isdir():
             cms_css.mkdir_p()
-            THEME_SASS_DIRECTORIES.append(cms_sass)
+            THEME_SASS_DIRECTORIES.append(("cms/static/sass", theme_dir / "cms" / "static" / "css"))
```
Nit: You can replace `theme_dir / "cms" / "static" / "css"` with `cms_css`, which is already defined a couple of lines above.

mtyaka commented on pavelib/assets.py:
```diff
@@ -80,12 +98,46 @@ def applicable_sass_directories(systems=None):
+        systems: A list of systems (defaults to all)
+
+    Returns:
+        A list of SASS directories to be compiled.
```
Nit: This should say something like "A list of SASS directories to be added to SASS lookup path."

mtyaka commented on pavelib/assets.py:
```diff
@@ -126,7 +178,7 @@ def register(self, observer):
         register files with observer
         """
-        for dirname in SASS_LOAD_PATHS + applicable_sass_directories():
+        for dirname in COMMON_SASS_LOOKUP_DIRECTORIES + sass_source_directories():
```
I think the watcher should watch all lookup directories. I would change this line to:

    for dirname in sass_lookup_directores() + sass_source_directories():

mtyaka replied:
Actually, since `sass_source_directores` returns a list of path pairs/tuples, this should probably be something like:

I think the watcher should watch all lookup directories. I would change this line to:

    for dirname in sass_lookup_directores() + [dir for dir, _ in sass_source_directories()]:

mtyaka commented on pavelib/assets.py:
```diff
@@ -20,20 +20,34 @@
+
 THEME_SASS_DIRECTORIES = []
-SASS_LOAD_PATHS = ['common/static', 'common/static/sass']
+THEME_SASS_LOOKUP_DIRECTORIES = []
```
I think we'll have to split `THEME_SASS_DIRECTORIES` into `THEME_LMS_SASS_DIRECTORIES` and `THEME_CMS_SASS_DIRECTORIES`; and `THEME_SASS_LOOKUP_DIRECTORIES` into `THEME_SASS_LMS_LOOKUP_DIRECTORIES` and `THEME_SASS_CMS_LOOKUP_DIRECTORIES`.

Currently it doesn't matter much because most of the time we compile all sass files in a single command (`systems` equals `ALL_SYSTEMS`), but if we decide to move more sass files to the `partials` directory, we will quickly start to get name clashes between partials in the lms partials folder and partials in the cms partials folder, so it will be important to compile lms and cms sass in two separate commands.

mtyaka commented on pavelib/assets.py:
```diff
@@ -240,17 +292,16 @@ def compile_sass(options):
-    system_sass_directories = applicable_sass_directories(systems)
-    all_sass_directories = applicable_sass_directories()
+    system_sass_directories = sass_source_directories(systems)
+    all_sass_directories = sass_lookup_directories()
```
I think this should still be scoped to current system: `sass_lookup_directories(system)`. I am not sure why it was always adding all paths (lms + cms) to the load path before, but that shouldn't be necessary. Any file that is shared between lms and cms sass should go into the common sass directory; we shouldn't be adding all directories to the load path here.

mtyaka replied:
Nit: Can we change the variable name from `all_sass_directories` to something like `system_sass_lookup_directories`?

mtyaka commented:
@saleem-latif Thanks for this, it looks great.

In addition to the comments I made inline:
- Can we move the `_variables.scss` file one level up - from `partials/base/_variables.scss` to `partials/_variables.scss`? I imagine almost every theme will want to override this file, so it would be nicer if theme authors didn't have to create a `base` subfolder in their theme's `partials` folder just to override `_variables.scss`.
- I would move *all* partial sass files to the partials directory. We allow theme authors to override any template and any static file, so I think it would make sense to let them override any sass file as well.

saleem-latif commented:
@mtyaka I tried moving `_variables.scss` to partials without base dir but then it would conflict with `_variables.scss` in `cms/statis/sass` dir. that is why I put it inside `partials/base` . but if we compile lms and cms's sass in separate commands then this problem will be resolved, as you mentioned in your previos comment, but I am not sure if how this approach will affect other things

benpatterson commented on pavelib/assets.py:
```diff
@@ -20,20 +20,34 @@
-    path("common/static/sass"),
+    (path("common/static/sass"), path("common/static/css"))
+]
+COMMON_SASS_LOOKUP_DIRECTORIES = [
```
I think we need more commentary in the code describing the function of the `lookup` directories.

benpatterson commented:
This PR needs tests.

benpatterson commented:
I'm a little concerned with how reliant this approach is to execution order. Can there be a way to do this that is more deterministic?

andy-armstrong commented on pavelib/assets.py:
```diff
@@ -240,17 +292,16 @@ def compile_sass(options):
+                    css_dir=css_dir or sass_dir.parent / "css",
                 ))
             else:
                 sh("rm -rf {css_dir}/*.css".format(css_dir=css_dir))
```
Doesn't this need to handle ``css_dir`` being null, like the ``dry_run`` case does?

andy-armstrong commented on pavelib/assets.py:
```diff
@@ -80,12 +98,46 @@ def applicable_sass_directories(systems=None):
+    applicable_directories.extend(COMMON_SASS_LOOKUP_DIRECTORIES)
+
+    # Put theme sass at the top so that theme directories have highest priority on sass file lookup.
+    applicable_directories.extend([sass_dir for sass_dir, _ in THEME_SASS_DIRECTORIES])
```
I expected the logic to only add the LMS theme directory when "lms" was being compiled, and the same for CMS. Is it necessary to add all the themes all the time? Can you maintain ``LMS_THEME_SASS_DIRECTORIES`` and for CMS too?

andy-armstrong commented on pavelib/assets.py:
```diff
@@ -262,7 +313,7 @@ def compile_sass(options):
         else:
             sass.compile(
                 dirname=(sass_dir, css_dir),
```
What happens here if ``css_dir`` is None?

andy-armstrong commented:
I agree with @benpatterson that this needs tests. Look at test_assets.py and extend it to demonstrate the behavior when a comprehensive theme is in play.

andy-armstrong commented:
I'm a little unsure about the idea of overriding partial files. It seems brittle to me, in that it is depending upon the naming conventions used to break up the SASS. It means that going forward, platform developers can't rename anything or change load order without the risk of breaking themes. It also means that the theme author has to know exactly how the platform author laid out all of their partials. If we do go down this path then we will have to be very careful to document what can be overridden by theme authors, and what therefore can't be changed by platform developers. We have essentially made the files on the disk into a contract that can't be changed.

There is also a risk of accidental overrides if a theme author has a new SASS partial, and then later the platform author happens to use the same filename. In this situation, the platform author won't know that they've broken the theme.

Having said this, it does seem very powerful too. It makes it much easier to change theme colors, fonts etc in a comprehensive way, without having to manually override every single CSS file with a theme version. If we are very explicit about the files that a theme author is allowed to override then maybe that's okay. One option worth considering is to have a named directory for the partials that are allowed to be overridden, to avoid any risk of accidental collisions.

andy-armstrong commented:
Thinking this through some more, I'm confused as to how this will work (this could well be because I'm not understanding things correctly). Say you have three installed themes: red, green and blue. Each one of these overrides variables.scss to define the brand colors to be appropriate to their theme. In order to support multi-tenancy, the result we want is that each theme directory has an lms-main-ltr.css (for example) that is compiled with its brand colors applied. When a request asks for the blue theme, the static file loader will look for the CSS file first in the blue theme directory and will pick it up if it is there. If it isn't, then it will instead load the default version of the CSS file.

If I'm understanding this PR correctly, what it does is put all three of these theme directories on the lookup path at once. This means that only the first listed theme will get to override variables.scss, and that this theme's values will be used to generate the code lms-main-ltr.css. This won't support multi-tenancy where we can dynamically switch themes per request.

Is my understanding correct here? To reiterate @benpatterson's point, it is crucial that we have tests that demonstrate that the desired use cases will work. Similarly, I recommend that we implement a URL parameter so that developers can easily switch themes, like we can with ``?preview-lang=foo``. If we can do ``?preview-theme=red`` and then switch to ``?preview-theme=blue`` that would allow us to make sure that the theming really is per request.

mtyaka commented:
@andy-armstrong This PR only deals with SASS compilation, not with django static asset lookup path. It currently assumes a single theme, but it wouldn't be hard to extend this to support multiple themes; sass compilation would just have to be run separately for each installed theme.

Multi-tenancy and multi-theme support are being worked on in a separate PR: https://github.com/edx/edx-platform/pull/11480

mtyaka commented:
@andy-armstrong I understand your concerns about themes being brittle if we allow overriding arbitrary sass partials. The same could be said of templates and static files (images, etc), though. We already support overriding arbitrary django/mako templates and static files from a theme, but that doesn't mean we can't change or rename any template in edx-platform. As a theme author, I understand overriding templates is brittle, but I still do it sometimes when there's no better solution, and I am grateful I can do that in a custom theme instead of having to change templates directly in a fork of edx-platform.

Themes almost always have to be adapted for new major releases of edx-platform due to changes in HTML structures and CSS rules, so theme authors expect to have to do some work when upgrading.

Still, I don't think overriding partials other than a few selected and documented ones (ie. `_variables.scss`) should be considered good practice, but I would really like to have the ability to do that.

andy-armstrong commented:
@mtyaka Thanks for the well thought out replies. 

It is good to hear that you don't think it would be hard to add support for multiple themes. To my mind, I wouldn't want to merge this without it, because we know that's what we need. I think the key is that the theme's SASS overrides should only apply to the theme's own CSS files, which as you suggest seems that there should be a separate compilation step performed per theme.

Regarding overrides, I think we're in agreement. Overriding most partials will be risky and error prone, so we should just make it clear which ones make sense, i.e. ``_variables.scss`` and a handful of others.

I'm on vacation all next week, so if possible it would be ideal to hold off merging until I get back. If you need this sooner then I recommend getting @AlasdairSwan's opinion. Thanks.

saleem-latif commented:
@benpatterson , @mtyaka , @andy-armstrong  I have incorporated most of the feedback and updated code to address the concerns mentioned in the feedback and added unit tests for the changes set, kindly take a look and let me know if I missed something.

Please let me know which files to put into partials from `sass/base`.

@andy-armstrong current implementation allows only one theme per OpenEdx installation and the original goal of this PR was to allow theme authors to override sass files the way they override templates/static files. Once we allow multiple themes into edx we would also need to update sass compilation process to support multi-tenancy. One implementation that comes to mind for multi-tenancy is we make `COMPREHENSIVE_THEME_DIR` a list, instead of a string, of all themes and in `compile_sass` we can start subtasks for the sass compilation of each individual theme.

mattdrayer commented:
@AlasdairSwan, it would be great if you could take a look at this PR -- thanks!

mattdrayer commented:
jenkins run python

mattdrayer commented:
jenkins run lettuce

benpatterson commented:
I'm about half-way through a pass on this. However, this still seems fairly dependent on the order of operations. Is there a smarter way to do it?

mattdrayer commented:
@benpatterson, it appears that @saleem-latif has updated this PR with a new commit -- can you take another look and let us know what else you'd like to see?

benpatterson commented:
@saleem-latif did you make additional changes? It looks like a squash and/or rebase to me...

saleem-latif commented:
@benpatterson yes I made the changes that you suggested in the feedback.

saleem-latif commented:
jenkins run lettuce

saleem-latif commented:
@benpatterson I have divided test cases into multiple tests, kindly take a look.

andy-armstrong commented on openedx/core/djangoapps/theming/tests/test_theme_styling.py:
```diff
@@ -0,0 +1,265 @@
+        """
+        Disable comprehensive theme and clear changes made for comprehensive themes.
+        """
+        patch("pavelib.assets.Env.env_tokens", {'COMPREHENSIVE_THEME_DIR': ""})
```
Why is this line necessary when the env_tokens is only updated within a ``with`` clause that should clean up?

andy-armstrong commented on openedx/core/djangoapps/theming/tests/test_theme_styling.py:
```diff
@@ -0,0 +1,265 @@
+        clear_theme_sass_dirs()
+
+    @with_comprehensive_theme(settings.TEST_THEME)
+    def test_green_footer(self):
```
Nit: why is this called ``test_green_footer``?

andy-armstrong commented on openedx/core/djangoapps/theming/tests/test_theme_styling.py:
```diff
@@ -0,0 +1,265 @@
+        Disable comprehensive theme and clear changes made for comprehensive themes.
+        """
+        patch("pavelib.assets.Env.env_tokens", {'COMPREHENSIVE_THEME_DIR': ""})
+        clear_theme_sass_dirs()
```
I wonder if you should call ``compile_sass`` again to regenerate the test files. If you don't, then someone who runs this test will then have an incorrect css file for any subsequent tests that they run.

andy-armstrong replied:
For example, Bok Choy tests use the same CSS files from ``test_root/staticfiles``. Maybe the kind of differences that are made by the theming overrides won't cause any trouble, but it seems risky.

andy-armstrong commented:
@saleem-latif Thanks for addressing all of my feedback. The code looks very clear now, and the tests are excellent.

Unfortunately I'm still uncomfortable with the approach, i.e. that the configured theme changes the way that the non-themed CSS files are generated. This seems like it will be confusing to folks who try to change their theme but don't think to regenerate their CSS. It will be important to document this clearly in the comprehensive theming documentation. My bigger concern is that this approach doesn't work for multi-tenancy, and we know that's where we want to go. I looked at the multi-tenancy PR (https://github.com/edx/edx-platform/pull/11480) but that doesn't seem to address SASS compilation as yet.

I don't see that there's enough benefit to the theme developer with this change, when they will have to change everything they do when we support multi-tenancy. Is the thinking that this approach will continue to work for folks who don't need multi-tenancy even when we add full support for it for ourselves? How are you imagining that this will work? I'd like to see a clear architectural design for multi-tenant theming and understand how this work fits into it before approving this PR.

@mattdrayer can we discuss this the next time you're in the office so that we can both be on the same page. FYI @efagin @nedbat.

caesar2164 commented:
@andy-armstrong & at devs: wouldn't it be better to allow overriding of just the base colors (i.e. $black, $red, $blue, etc...) and then let all the other vars just automagically generated themselves?

caesar2164 commented:
@mattdrayer & @saleem-latif, so you guys get a ping about ^

mattdrayer commented:
@andy-armstrong PR #11480 is an older attempt at multi-tenant theming that has since been closed -- instead please reference #11613

saleem-latif commented:
Closing this PR as the development of the feature has shifted to #11613
